### PR TITLE
Add a metric batcher which preallocates batch sizes.

### PR DESCRIPTION
**Pull request by bogdandrutu**

### CODE REVIEW

File: metrics_proto.go [hunk 1]
```diff
@@ -343,9 +234,9 @@ func resourcepbToResource(rsc *resourcepb.Resource) *resource.Resource {
 
 // protoMetricToTimeSeries converts a metric into a Stackdriver Monitoring v3 API CreateTimeSeriesRequest
 // but it doesn't invoke any remote API.
-func (se *statsExporter) protoMetricToTimeSeries(ctx context.Context, mappedRsc *monitoredrespb.MonitoredResource, metric *metricspb.Metric, additionalLabels map[string]labelValue) ([]*monitoringpb.TimeSeries, error) {
+func (se *statsExporter) protoMetricToTimeSeries(ctx context.Context, mappedRsc *monitoredrespb.MonitoredResource, metric *metricspb.Metric, mb *metricsBatcher) {
 	if metric == nil || metric.MetricDescriptor == nil {
-		return nil, errNilMetricOrMetricDescriptor
+		mb.recordDroppedTimeseries(len(metric.GetTimeseries()), errNilMetricOrMetricDescriptor)
```
Comment (songy23) [inline metrics_proto.go]: `metric.GetTimeseries()` may produce a nil pointer error.

File: metrics_batcher.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		mb.allReqs = append(mb.allReqs, &monitoringpb.CreateTimeSeriesRequest{
+			Name:       monitoring.MetricProjectPath(mb.projectID),
+			TimeSeries: mb.allTss,
+		})
+	}
+
+	// Send create time series requests to Stackdriver.
+	for _, req := range mb.allReqs {
+		if err := createTimeSeries(ctx, mc, req); err != nil {
+			mb.droppedTimeSeries += len(req.TimeSeries)
```
Comment (songy23) [inline metrics_batcher.go]: Use `mb.recordDroppedTimeseries` instead?

File: metrics_proto.go (hunk 1 above)
Comment (bogdandrutu) [reply metrics_proto.go]: No:
func (m *Metric) GetTimeseries() []*TimeSeries {
	if m != nil {
		return m.Timeseries
	}
	return nil
}

File: metrics_batcher.go (hunk 2 above)
Comment (bogdandrutu) [reply metrics_batcher.go]: Fixed the code now, I think this makes more sense.

File: metrics_batcher.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+func (mb *metricsBatcher) recordDroppedTimeseries(numTimeSeries int, err error) {
+	mb.droppedTimeSeries += numTimeSeries
+	mb.receivedTimeSeries += numTimeSeries
+	mb.allErrs = append(mb.allErrs, err)
+}
+
+func (mb *metricsBatcher) addTimeSeries(ts *monitoringpb.TimeSeries) {
+	mb.receivedTimeSeries++
+	mb.allTss = append(mb.allTss, ts)
+	if len(mb.allReqs) == maxTimeSeriesPerUpload {
```
Comment (rghetia) [inline metrics_batcher.go]: should this be len(mb.allTss)?

File: metrics_batcher.go (hunk 2 above)
Comment (rghetia) [reply metrics_batcher.go]: recordDroppedTimeseries will double count receivedTimeSeries, isn't it?

File: metrics_batcher.go (hunk 3 above)
Comment (bogdandrutu) [reply metrics_batcher.go]: Done.

**Comment by bogdandrutu** [comment]
PTAL. Added a method that returns number of dropped timeseries. So we can use the helper for metrics recording in https://github.com/open-telemetry/opentelemetry-service/pull/315

**Comment by rghetia** [comment]
LGTM

**Pull request merged**